services/kaze/controllers/application.go

```diff
@@ -242,15 +245,23 @@ func FetchMetrics(c *gin.Context) {
 		}
 	}
 
-	metrics := mongo.FetchContainerMetrics(types.M{
-		mongo.NameKey: appName,
-		mongo.TimestampKey: types.M{
-			"$gte": time.Now().Unix() - timeSpan,
-		},
-	})
-
-	c.JSON(200, gin.H{
-		"success": true,
-		"data":    metrics,
+	chanStream := make(chan bool)
+	go func() {
+		defer close(chanStream)
+		chanStream <- true
+		time.Sleep(2 * configs.ServiceConfig.Mizu.MetricsInterval)
+	}()
+	c.Stream(func(w io.Writer) bool {
+		if _, ok := <-chanStream; ok {
+			metrics := mongo.FetchContainerMetrics(types.M{
+				mongo.NameKey: appName,
+				mongo.TimestampKey: types.M{
+					"$gte": time.Now().Unix() - timeSpan,
+				},
+			})
+			c.SSEvent("metrics", metrics)
+			return true
+		}
+		return false
```

Comment:
the overall function will be simply
```
	c.Stream(func(w io.Writer) bool {
			metrics := mongo.FetchContainerMetrics(types.M{
				mongo.NameKey: appName,
				mongo.TimestampKey: types.M{
					"$gte": time.Now().Unix() - configs.ServiceConfig.Mizu.MetricsInterval * time.Second,
				},
			})
			c.SSEvent("metrics", metrics)
                        time.Sleep(configs.ServiceConfig.Mizu.MetricsInterval * time.Second)
			return true
		}
```

Reply:
How is using a channel any different than the above function ^ ?

Reply:
Well, as was discussed over slack, there was unknown `i/o timeout` error. I still am not sure about the exact reason but here what I infer from the success:
* The original `time.Sleep()` was forcing the stream to halt somehow, but SSEvent was not prepared for discrete streams. 
* So by using this channel mutex triggered SSEvent, it is suppressed during the halt duration. And hence no timeout errors. I tried to mimic some of the errors by manipulating the goroutine and this inference feels rational.
* gin's SSEvent implementation is buggy and we might need a replacement later.